File: src/main/java/dev/lonami/uniffidl/UdlFindUsagesProvider.java
package dev.lonami.uniffidl;

import com.intellij.lang.cacheBuilder.DefaultWordsScanner;
import com.intellij.lang.cacheBuilder.WordsScanner;
import com.intellij.lang.findUsages.FindUsagesProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import com.intellij.psi.tree.TokenSet;
import dev.lonami.uniffidl.psi.*;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class UdlFindUsagesProvider implements FindUsagesProvider {
    @Override
    public @Nullable WordsScanner getWordsScanner() {
        return new DefaultWordsScanner(new UdlLexerAdapter(),
                UdlTokenSets.IDENTIFIERS,
                UdlTokenSets.COMMENTS,
                TokenSet.EMPTY);
    }

    @Override
    public boolean canFindUsagesFor(@NotNull PsiElement psiElement) {
        return psiElement instanceof PsiNamedElement;
    }

    @Override
    public @Nullable @NonNls String getHelpId(@NotNull PsiElement psiElement) {
        return null;
    }

    @Override
    public @Nls @NotNull String getType(@NotNull PsiElement element) {
        if (element instanceof UdlTypedef) {
            return "typedef";
        } else if (element instanceof UdlDictionary) {
            return "dictionary";
        } else if (element instanceof UdlEnum) {
            return "enum";
        }
        return "";
    }

    @Override
    public @Nls @NotNull String getDescriptiveName(@NotNull PsiElement element) {
        if (element instanceof UdlDefinition) {
            String name = ((UdlDefinition) element).getName();
            return name != null ? name : "";
        }
        return "";
    }

    @Override
    public @Nls @NotNull String getNodeText(@NotNull PsiElement element, boolean useFullName) {
        if (element instanceof UdlDefinition) {
            String text = element.getText();
            return text != null ? text : "";
        }
        return "";
    }
}

File: src/main/kotlin/dev/lonami/uniffidl/UdlFindUsagesProvider.kt
package dev.lonami.uniffidl

import com.intellij.lang.cacheBuilder.DefaultWordsScanner
import com.intellij.lang.cacheBuilder.WordsScanner
import com.intellij.lang.findUsages.FindUsagesProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.tree.TokenSet
import dev.lonami.uniffidl.psi.*
import org.jetbrains.annotations.Nls
import org.jetbrains.annotations.NonNls

class UdlFindUsagesProvider : FindUsagesProvider {
    override fun getWordsScanner(): WordsScanner {
        return DefaultWordsScanner(
            UdlLexerAdapter(),
            UdlTokenSets.IDENTIFIERS,
            UdlTokenSets.COMMENTS,
            TokenSet.EMPTY
        )
    }

    override fun canFindUsagesFor(psiElement: PsiElement): Boolean {
        return psiElement is PsiNamedElement
    }

    override fun getHelpId(psiElement: PsiElement): @NonNls String? = null

    override fun getType(element: PsiElement): @Nls String {
        return when (element) {
            is UdlTypedef -> "typedef"
            is UdlDictionary -> "dictionary"
            is UdlEnum -> "enum"
            else -> ""
        }
    }

    override fun getDescriptiveName(element: PsiElement): @Nls String {
        if (element is UdlDefinition) {
            val name = element.name
            return name ?: ""
        }
        return ""
    }

    override fun getNodeText(element: PsiElement, useFullName: Boolean): @Nls String {
        if (element is UdlDefinition) {
            val text = element.text
            return text ?: ""
        }
        return ""
    }
}
